### ISSUE TITLE
No batch_matmul operation on TensorFlow r0.12

### ISSUE THREAD

**Issue by Itaydal** [fchollet/keras#4864]
I have just updated my TensorFlow to version r0.12
It looks like they are no longer supporting 'batch_matmul' operation 

Specific error:

  File "/home/itay/anaconda3/lib/python3.5/site-packages/keras/backend/tensorflow_backend.py", line 871, in batch_dot
    out = tf.batch_matmul(x, y, adj_a=adj_x, adj_b=adj_y)

AttributeError: module 'tensorflow' has no attribute 'batch_matmul'

What can I do to solve this?

**Comment by patyork** [comment]
You should upgrade Keras from github (instead of from Pip/anaconda) and then ensure TensorFlow was installed without issue; [TensorFlow definitely still supports `batch_matmul`](https://www.tensorflow.org/api_docs/python/math_ops/matrix_math_functions#batch_matmul).

**Comment by siebertlooije** [comment]
Got the same issue as @Itaydal. After updating to 0.12.1 the error occured: 

  File "/home/siebert/.local/lib/python3.5/site-packages/keras/engine/topology.py", line 528, in __call__
    return self.call(x, mask)
  File "/home/siebert/.local/lib/python3.5/site-packages/keras/layers/local.py", line 400, in call
    output = K.batch_dot(x_aggregate, self.W)
  File "/home/siebert/.local/lib/python3.5/site-packages/keras/backend/tensorflow_backend.py", line 436, in batch_dot
    out = tf.batch_matmul(x, y, adj_x=adj_x, adj_y=adj_y)
AttributeError: module 'tensorflow' has no attribute 'batch_matmul'

I already pulled the latest version of keras and installed it but still got the same error.

Any has an idea ?

**Comment by Itaydal** [comment]
@patyork 
Looks like it was removed after all 

[https://github.com/tensorflow/tensorflow/issues/6560](url)

[https://github.com/tensorflow/tensorflow/commit/3a88ec0b8e5a9b8f3f3efc4d9a8f88d47b3f4042](url)

What can we do about it?

**Comment by patyork** [comment]
It looks like the call to `tf.batch_matmul()` in the tensorflow_backend.py can just be changed to `tf.matmul()`.

You could implement the fix and do a Pull Request.

**Comment by Itaydal** [comment]
@patyork 
But than wont it break support for older versions of TensorFlow?

P.S
Could you please refer me to some guid that explains how do I do Pull request for Keras?

**Comment by Itaydal** [comment]
@fchollet Could you help please?
Currently Keras does not support TF 0.12

**Comment by patyork** [comment]
@Itaydal You can check out the [Contributing page](https://github.com/fchollet/keras/blob/master/CONTRIBUTING.md#pull-requests) for more info on pull requests.

In terms of breaking changes, you can probably just add a new `except AttributeError` [here](https://github.com/fchollet/keras/blob/master/keras/backend/tensorflow_backend.py#L872), or you can wrap that try-catch in an if statement to check if TF has batch_matmul (meaning it can be used) else use the new `matmul` section; you can check out this [snippet](https://github.com/fchollet/keras/blob/master/keras/backend/tensorflow_backend.py#L258) for an example.

I think there were a couple other people working on TF.12 fixes and changes, but I'm not sure what the status is.

**Comment by drspiffy** [comment]
It seems like this ought to be possible to solve using tf.matmul inside tf.scan. I'm going to try to code this out because I write an attention mechanism that uses batch_dot which now I can only run with Theano.

**Comment by patyork** [comment]
For TF.12 this has been solved. If there are issues with TF1.0, then this should be a separate issue.

Just for posterity's sake, I will say, that if you ever find yourself saying that "in a loop" is better than "in parallel" you should reread the documentation.
